pkg/util/alertmanager/alertmanager.go

```diff
@@ -0,0 +1,84 @@
+package alertmanager
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/prometheus/common/model"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/rest"
+
+	"github.com/Azure/ARO-RP/pkg/util/portforward"
+)
+
+type AlertManager interface {
+	FetchPrometheusAlerts(ctx context.Context) ([]model.Alert, error)
+}
+
+type alertManager struct {
+	log        *logrus.Entry
+	restConfig *rest.Config
+}
+
+func NewAlertManager(c *rest.Config, log *logrus.Entry) AlertManager {
+	return &alertManager{log, c}
+}
+
+func (a *alertManager) FetchPrometheusAlerts(ctx context.Context) ([]model.Alert, error) {
+	var resp *http.Response
+	var err error
+
+	for i := 0; i < 3; i++ {
+		hc := &http.Client{
+			Transport: &http.Transport{
+				DialContext: func(ctx context.Context, network, address string) (net.Conn, error) {
+					_, port, err := net.SplitHostPort(address)
+					if err != nil {
+						return nil, err
+					}
+					return portforward.DialContext(ctx, a.log, a.restConfig, "openshift-monitoring", fmt.Sprintf("alertmanager-main-%d", i), port)
+				},
+				// HACK: without this, keepalive connections don't get closed,
+				// resulting in excessive open TCP connections, lots of
+				// goroutines not exiting and memory not being freed.
+				// TODO: consider persisting hc between calls to Monitor().  If
+				// this is done, take care in the future to call
+				// hc.CloseIdleConnections() when finally disposing of an hc.
+				DisableKeepAlives: true,
+			},
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, "http://alertmanager-main.openshift-monitoring.svc:9093/api/v2/alerts", nil)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err = hc.Do(req)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code %d", resp.StatusCode)
+	}
+
+	var alerts []model.Alert
+	err = json.NewDecoder(resp.Body).Decode(&alerts)
+	if err != nil {
+		return nil, err
+	}
+
+	return alerts, nil
+}
```

Comment:
please split this function into smaller ones and add some testing for it. Since we mock the whole function, we need to make sure it is doing what we are expecting it to do

Reply:
also, is the comment from you or did you get it from somewhere else?

Reply:
I reused the some bits from `pkg/monitor/cluster` package

Reply:
Got it, shall work on the breaking down the function.

Reply:
Done with splitting the function and added test for it